### PR TITLE
Update the Gaudi trainer with transformers 4.45.2

### PR DESCRIPTION
**Pull request by yafshar**
# What does this PR do?

 Update the Gaudi trainer with transformers 4.45.2

- Add the description
- remove _is_peft_model function and import it from transformers
- remove the unnecessary import datasets and keep the first one
- Update the _get_train_sampler, by replacing computed var num_samples
- Update the _inner_training_loop
  - num_update_steps_per_epoch used when possible
  - unused line was removed
  - add a new _should_compute_grad_norm var to remove extra conditional check
- Update the _load_best_model, enabled the if self.is_deepspeed_enabled
- Update _maybe_log_save_evaluate, _grad_norm.item() -> _grad_norm.detach.item()
- Remove the _save_checkpoint, exactly the same as transformers
- Update autocast_smart_context_manager, updateing the old interface
- Update save_model, add accelerate version check
- Update evaluation_loop
  - introduce a new var _should_update_inputs  for llama, qwen2, starcoder2 and gemma, to avoid double if condition in the loop
  - set the logits_dtype to float32 before the loop
  - re-order the conditionals for losses, logits, labels and match with
    transformers
- Update _inner_training_loop
  - introduce a new var _should_update_inputs for llama, qwen2, starcoder2 and gemma, to avoid double if condition in the loop

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by yafshar** [comment]
The env for all tests,

```sh
export RUN_SLOW=true
export GAUDI2_CI=1
```

Some tests which are finished

```sh
- slow_tests_fsdp -> 2 passed in 689.82s (0:11:29) | the same accuracy and loss on both tests with main

- test_trainer_distributed -> 2 passed, 6 warnings in 21.19s | the same as main

- test_trainer -> 75 passed, 8 skipped, 37 warnings in 55.23s | the same as main

- test_trainer_seq2seq > 2 passed, 6 warnings in 21.19s | the same as main
```

Other tests are in progress and will be updated here

**Comment by yafshar** [comment]
The env for all tests,

```sh
export RUN_SLOW=true
export GAUDI2_CI=1
```

test_examples  finished

```sh
- test_examples -> 11 failed, 50 passed, 2 warnings in 21761.68s (6:02:41) | 13 failed, 48 passed, 2 warnings in 23254.39s (6:27:34) on main
```

- main branch
```sh 
=========================== short test summary info ============================
FAILED tests/test_examples.py::MultiCardQuestionAnsweringExampleTester::test_run_qa_albert-large-v2_multi_card
FAILED tests/test_examples.py::MultiCardQuestionAnsweringExampleTester::test_run_qa_roberta-large_multi_card
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingExampleTester::test_run_clm_gpt2_multi_card
FAILED tests/test_examples.py::DeepspeedCausalLanguageModelingExampleTester::test_run_clm_CodeLlama-13b-Instruct-hf_deepspeed
FAILED tests/test_examples.py::MultiCardSpeechRecognitionExampleTester::test_run_speech_recognition_ctc_wav2vec2-large-lv60_multi_card
FAILED tests/test_examples.py::MultiCardSummarizationExampleTester::test_run_summarization_t5-small_multi_card
FAILED tests/test_examples.py::DeepspeedSummarizationExampleTester::test_run_summarization_flan-t5-xxl_deepspeed
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingLORAExampleTester2::test_run_lora_clm_falcon-40b_multi_card
FAILED tests/test_examples.py::MultiCardSeq2SeqSpeechRecognitionExampleTester::test_run_speech_recognition_seq2seq_whisper-small_multi_card
FAILED tests/test_examples.py::DeepspeedSFTExampleTester::test_sft_Qwen2-72B_deepspeed
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingPrefixTuningExampleTester::test_run_prompt_tuning_clm_llama-7b_multi_card
FAILED tests/test_examples.py::MultiCardMultiTastPromptPeftExampleTester::test_run_multitask_prompt_tuning_t5-small_multi_card
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingVeraExampleTester::test_run_lora_clm_llama-7b_multi_card
=========== 13 failed, 48 passed, 2 warnings in 23254.39s (6:27:34) ============
```

Current PR
```sh
=========================== short test summary info ============================
FAILED tests/test_examples.py::MultiCardQuestionAnsweringExampleTester::test_run_qa_roberta-large_multi_card
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingExampleTester::test_run_clm_gpt2_multi_card
FAILED tests/test_examples.py::DeepspeedCausalLanguageModelingExampleTester::test_run_clm_CodeLlama-13b-Instruct-hf_deepspeed
FAILED tests/test_examples.py::MultiCardSummarizationExampleTester::test_run_summarization_t5-small_multi_card
FAILED tests/test_examples.py::DeepspeedSummarizationExampleTester::test_run_summarization_flan-t5-xxl_deepspeed
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingLORAExampleTester2::test_run_lora_clm_falcon-40b_multi_card
FAILED tests/test_examples.py::MultiCardSeq2SeqSpeechRecognitionExampleTester::test_run_speech_recognition_seq2seq_whisper-small_multi_card
FAILED tests/test_examples.py::DeepspeedSFTExampleTester::test_sft_Qwen2-72B_deepspeed
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingPrefixTuningExampleTester::test_run_prompt_tuning_clm_llama-7b_multi_card
FAILED tests/test_examples.py::MultiCardMultiTastPromptPeftExampleTester::test_run_multitask_prompt_tuning_t5-small_multi_card
FAILED tests/test_examples.py::MultiCardCausalLanguageModelingVeraExampleTester::test_run_lora_clm_llama-7b_multi_card
=========== 11 failed, 50 passed, 2 warnings in 21761.68s (6:02:41) ============
```

**Comment by yafshar** [comment]
@regisss would you please check this PR, if we want to move towards v4.46.0, this PR will help.

**Comment by emascarenhas** [comment]
Please make the title more descriptive to say what PR is doing.

File: optimum/habana/transformers/trainer.py [hunk 1]
```diff
@@ -867,6 +846,28 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args, use_reentrant: Optio
         self._globalstep_last_logged = self.state.global_step
         self._zero_model_grad(model)
         _grad_norm: Optional[float] = None
+        _should_compute_grad_norm: bool = not self.accelerator.distributed_type == GaudiDistributedType.DEEPSPEED and (
+            # Gradient clipping
+            args.max_grad_norm is not None and args.max_grad_norm > 0
+        )
+
+        # attn_softmax_bf16 and use_flash_attention are enabled only for llama, qwen2, starcoder2 and gemma
```
Comment (emascarenhas) [inline optimum/habana/transformers/trainer.py:854]: Code is ok as is, however, would it be better to put this in a function which returns the setting for _should_update_inputs ? the list of models may expand in the future, so may be better to set it somewhere else?

File: optimum/habana/transformers/trainer.py [hunk 2]
```diff
@@ -1318,72 +1309,6 @@ def _load_rng_state(self, checkpoint):
                         "\nThis won't yield the same results as if the training had not been interrupted."
                     )
 
-    def _save_checkpoint(self, model, trial, metrics=None):
```
Comment (emascarenhas) [inline optimum/habana/transformers/trainer.py:1321]: There is a call to _save_checkpoint on line 1267. Wondering if the function got moved elsewhere?

File: optimum/habana/transformers/trainer.py [hunk 3]
```diff
@@ -1637,7 +1562,9 @@ def save_model(self, output_dir: Optional[str] = None, _internal_call: bool = Fa
             output_dir = self.args.output_dir
 
         if self.is_fsdp_enabled:
-            if "FULL_STATE_DICT" in str(self.accelerator.state.fsdp_plugin.state_dict_type):
+            if ("FULL_STATE_DICT" in str(self.accelerator.state.fsdp_plugin.state_dict_type)) and (
+                is_accelerate_available("0.24.2")
+            ):
```
Comment (emascarenhas) [inline optimum/habana/transformers/trainer.py:1567]: Is there a better way to do this than hardcode the version here? It will no doubt need to be changed in the future.

File: optimum/habana/transformers/trainer.py [hunk 4]
```diff
@@ -1872,6 +1799,26 @@ def evaluation_loop(
         # Will be useful when we have an iterable dataset so don't know its length.
         observed_num_examples = 0
 
+        # attn_softmax_bf16 and use_flash_attention are enabled only for llama, qwen2, starcoder2 and gemma
+        _should_update_inputs: bool = (getattr(self.model, "generation_config", None) is not None) and (
+            self.model.config.model_type in ["llama", "qwen2", "starcoder2", "gemma"]
+        )
+
+        if _should_update_inputs:
```
Comment (emascarenhas) [inline optimum/habana/transformers/trainer.py:1807]: Some of this code can be refactored, see other comment above about turning it into a function.

**Review by emascarenhas — COMMENTED** [review]
@yafshar , Some comments added for code changes. Thanks for doing the port.

File: optimum/habana/transformers/trainer.py (hunk 2 above)
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1321]: The function is still in use and remains unchanged from the superclass `Trainer` in [`transformers`](https://github.com/huggingface/transformers/blob/53fad641cfdb5105e2470bcf3ef17ea8e25cc300/src/transformers/trainer.py#L2995) . Therefore, there is no need to rewrite it in this class.

File: optimum/habana/transformers/trainer.py (hunk 1 above)
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:888]: Thank you for your feedback. I agree that encapsulating the logic in a function could improve maintainability, especially as the list of models might expand in the future. I’ve refactored the code accordingly

File: optimum/habana/transformers/trainer.py (hunk 3 above)
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:1587]: I agree, but it is following the upstream code in [transformers](https://github.com/huggingface/transformers/blob/53fad641cfdb5105e2470bcf3ef17ea8e25cc300/src/transformers/trainer.py#L3600) and my goal is to make the upgrade easier for the next version 4.46.1 which has lots of changes

File: optimum/habana/transformers/trainer.py [hunk 5]
```diff
@@ -116,52 +124,58 @@
...
+            inputs_update["flash_attention_causal_mask"] = True
+        should_update_inputs = len(inputs_update) > 0
 
-            classes_to_check = (*classes_to_check, PeftMixedModel)
-        return isinstance(model, classes_to_check)
-    return False
+    return should_update_inputs, inputs_update
 
 
 if TYPE_CHECKING:
```
Comment (emascarenhas) [inline optimum/habana/transformers/trainer.py:174]: It looks like the transformers code still has this code under TYPE_CHECKING.

if TYPE_CHECKING:
    import optuna

    if is_datasets_available():
        import datasets

**Comment by emascarenhas** [comment]
@yafshar , Please redo make style, and a test run if you haven't already done so and confirm that no new test failures are seen. Thanks.

File: optimum/habana/transformers/trainer.py (hunk 5 above)
Comment (yafshar) [reply optimum/habana/transformers/trainer.py:174]: Yes, you are right. I have removed datasets in TYPE_CHECKING, since for transfomers it is optional but for optimum-habana it is a requirement, is imported above, thus there is no need to do it again.

**Comment by yafshar** [comment]
This PR,

```sh
>>>python -m pytest tests/test_gaudi_configuration.py tests/test_trainer_distributed.py tests/test_trainer.py tests/test_trainer_seq2seq.py -v -s
81 passed, 8 skipped, 140 warnings in 139.73s (0:02:19)

>>> python -m pytest tests/test_examples.py -v -s -k single_card
12 passed, 50 deselected in 2917.12s (0:48:37)

>>> python -m pytest tests/test_fsdp_examples.py -v -s
2 passed in 799.36s (0:13:19)

>>>  python -m pytest tests/test_examples.py -v -s -k multi_card
10 failed, 34 passed, 18 deselected, 2 warnings in 17206.15s (4:46:46)

>>> python -m pytest tests/test_examples.py -v -s -k deepspeed
3 failed, 3 passed, 56 deselected in 5636.71s (1:33:56)
```


main
```sh
>>>python -m pytest tests/test_gaudi_configuration.py tests/test_trainer_distributed.py tests/test_trainer.py tests/test_trainer_seq2seq.py -v -s
81 passed, 8 skipped, 140 warnings in 181.19s (0:03:01)

>>> python -m pytest tests/test_examples.py -v -s -k single_card
12 passed, 50 deselected in 2805.12s (0:46:45)

>>> python -m pytest tests/test_fsdp_examples.py -v -s
2 passed in 804.29s (0:13:24)

>>>  python -m pytest tests/test_examples.py -v -s -k multi_card
10 failed, 34 passed, 18 deselected, 2 warnings in 16397.55s (4:33:17)

>>> python -m pytest tests/test_examples.py -v -s -k deepspeed
3 failed, 3 passed, 56 deselected in 4832.87s (1:20:32)
```

**Comment by emascarenhas** [comment]
@libinta , Please add run-test and other labels to allow this to merge into OH.
@regiss, This is ready for review and merge.

**Comment by yafshar** [comment]
> @yafshar , The test_examples.py times are longer than the previous version. Worth investigating?

Sure, these runs are on different machines, but sure, I will check and report back here

File: optimum/habana/transformers/trainer.py [hunk 6]
```diff
@@ -117,53 +125,69 @@
...
 if is_deepspeed_available():
     from accelerate.utils import DeepSpeedSchedulerWrapper
 
-if is_accelerate_available():
-    from accelerate.utils import (
-        load_fsdp_optimizer,
-        save_fsdp_optimizer,
-    )
+if is_accelerate_available("0.28.0"):
+    from accelerate.utils import DataLoaderConfiguration
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:139]: We will have Accelerate >= v0.33.0 (cf. https://github.com/huggingface/optimum-habana/blob/ba914504bd52b989d05a4e11916aa19b3ce176f4/setup.py#L35) so we can remove that

File: optimum/habana/transformers/trainer.py [hunk 7]
```diff
@@ -1152,22 +1172,22 @@ def _load_best_model(self):
...
+                load_module_strict=not _is_peft_model(self.model),
+            )
+        elif self.is_fsdp_enabled:
+            load_result = load_fsdp_model(
+                self.accelerator.state.fsdp_plugin,
+                self.accelerator,
+                model,
+                self.state.best_model_checkpoint,
+                **_get_fsdp_ckpt_kwargs(),
+            )
```
Comment (regisss) [inline optimum/habana/transformers/trainer.py:1189]: Not sure this works :grin: 
Have you tested it?

File: optimum/habana/transformers/trainer.py (hunk 3 above)
Comment (regisss) [reply optimum/habana/transformers/trainer.py:1600]: I think it's safe to remove the version check as a more recent Accelerate version is required anyway: https://github.com/huggingface/optimum-habana/blob/ba914504bd52b989d05a4e11916aa19b3ce176f4/setup.py#L35

**Review by regisss — APPROVED** [review]
Nice!
I'll run the nightly CI to see if there is no regression. Also, not sure the `_load_best_model` method works with DeepSpeed and FSDP.

**Comment by emascarenhas** [comment]
> Nice! I'll run the nightly CI to see if there is no regression. Also, not sure the `_load_best_model` method works with DeepSpeed and FSDP.

@regisss , Do you have results from CI to share, was there any performance regression?

**Comment by yafshar** [comment]
with the latest changes

main branch

```sh
>>>  RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_examples.py -v -s --token=$HUGGING_FACE_HUB_TOKEN -k MultiCardCausalLanguageModelingLORAExampleTester

2 passed, 51 deselected in 2873.00s (0:47:52)
```

The current PR
```sh
>>>  RUN_SLOW=true GAUDI2_CI=1 python -m pytest tests/test_examples.py -v -s --token=$HUGGING_FACE_HUB_TOKEN -k MultiCardCausalLanguageModelingLORAExampleTester

2 passed, 51 deselected in 2301.59s (0:38:21)
```
